source/time_tables/departure_boards.cpp

```diff
@@ -270,11 +346,75 @@ void departure_board(PbCreator& pb_creator, const std::string& request,
         map_route_stop_point[route_point] = stop_times;
     }
 
-    render(pb_creator, response_status, map_route_stop_point, handler.date_time, handler.max_datetime,
-              calendar_id, depth);
+    render(pb_creator,
+           response_status,
+           map_route_stop_point,
+           map_route_first_point,
+           map_route_last_point,
+           handler.date_time,
+           handler.max_datetime,
+           calendar_id, depth);
 
     pb_creator.make_paginate(total_result, start_page, count, std::max(pb_creator.departure_boards_size(),
                                                                        pb_creator.stop_schedules_size()));
 }
+
+DateTime first_or_last_request_datetime(const datetime_type type,
+                                        const routing::datetime_stop_time& first_stop_time,
+                                        const uint32_t opening_time)
+{
```

Comment:
After spending time with @xlqian, it seems that `first_stop_time` actually is a `next_stop_time` :)

Also, returning the 2 values for opening and closing datetimes at the same time would simplify the readability. Maybe having something like:

```c++
std::pair<> get_opening_and_closing_datetime(
	const DateTime next_stoptime_datetime,
	const uint32_t opening_time)
{
	
	auto opening_datetime = DateTimeUtils::date(next_stoptime_datetime);
	auto opening_datetime = closing_datetime;

	if(next_stoptime_datetime > opening_time) {
	    opening_datetime += 1;
	}

	return make_pair( 
				DateTimeUtils::set(opening_datetime, opening_time)-1, 
				DateTimeUtils::set(opening_datetime, opening_time)-1);
}
```

Reply:
This is not so easy, but you gave me a good way to refactor this part. In any case, the `std::pair` in return and *variable name* is a good idea.